File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/dialog/AbstractDialogFragment.java
package com.wa2c.android.medoly.plugin.action.lrclyrics.dialog;

import android.app.Activity;
import android.app.Dialog;
import android.app.DialogFragment;
import android.app.Fragment;
import android.content.DialogInterface;
import android.content.SharedPreferences;
import android.content.res.Configuration;
import android.os.Bundle;
import android.preference.PreferenceManager;

import com.wa2c.android.medoly.plugin.action.lrclyrics.R;
import com.wa2c.android.medoly.plugin.action.lrclyrics.util.AppUtils;

import java.util.HashMap;
import java.util.Map;

/**
 * Abstract dialog.
 */
public abstract class AbstractDialogFragment extends DialogFragment {

    /**
     * Exists dialog map.
     */
    private static final Map<String, DialogFragment> shownDialogMap = new HashMap<>();

    /** Activity。 */
    protected Activity context;
    /** Preference. */
    protected SharedPreferences preferences;



    /***
     * Show dialog.
     * @param activity Activity.
     */
    public void show(Activity activity) {
        if (activity == null) {
            AppUtils.showToast(getActivity(), R.string.error_dialog_dismissed);
            return;
        }

        String key = this.getClass().getName();
        if (shownDialogMap.containsKey(key) && shownDialogMap.get(key) != null) {
            shownDialogMap.get(key).dismiss();
        }

        super.show(activity.getFragmentManager(), key);
        shownDialogMap.put(key, this);
    }

    /**
     * Show dialog.
     * @param fragment Fragment.
     */
    public void show(Fragment fragment) {
        if (fragment == null) {
            AppUtils.showToast(getActivity(), R.string.error_dialog_dismissed);
            return;
        }

        String key = this.getClass().getName();
        if (shownDialogMap.containsKey(key) && shownDialogMap.get(key) != null) {
            shownDialogMap.get(key).dismiss();
        }

        super.show(fragment.getFragmentManager(), key);
        shownDialogMap.put(key, this);
    }



    /**
     * onCreateDialog.
     */
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        context = getActivity();
        preferences = PreferenceManager.getDefaultSharedPreferences(getActivity());
        return super.onCreateDialog(savedInstanceState);
    }

    /**
     * onStart.
     */
    @Override
    public void onStart() {
        super.onStart();
    }

    /**
     * onConfigurationChanged.
     */
    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        super.onConfigurationChanged(newConfig);

        Dialog d = getDialog();
        if (d != null) d.cancel(); // close
    }

    /**
     * onStop
     */
    @Override
    public void onStop() {
        super.onStop();
        Dialog d = getDialog();
        if (d != null) d.cancel();
    }

    /**
     * onDismiss
     */
    @Override
    public void onDismiss(DialogInterface dialog) {
        super.onDismiss(dialog);

        String key = this.getClass().getName();
        shownDialogMap.remove(key);
    }



    /** Click listener. */
    protected DialogInterface.OnClickListener clickListener = null;

    /**
     * Set click listener.
     * @param listener Click listener.
     */
    public void setClickListener(DialogInterface.OnClickListener listener) {
        this.clickListener = listener;
    }

    /**
     * Execute button click event.
     * @param which Clicked button.
     */
    protected void onClickButton(int which) {
        onClickButton(which, true);
    }

    /**
     * Execute button click event.
     * @param which Clicked button.
     * @param close true if closing dialog.
     */
    protected void onClickButton(int which, boolean close) {
        Dialog dialog = this.getDialog();
        if (dialog != null && clickListener != null) {
            clickListener.onClick(dialog, which);
            if (close) dialog.dismiss();
        } else {
            if (dialog != null && close) dialog.cancel();
        }
    }

}

File: app/src/main/java/com/wa2c/android/medoly/plugin/action/lrclyrics/dialog/AbstractDialogFragment.kt
package com.wa2c.android.medoly.plugin.action.lrclyrics.dialog

import android.app.Dialog
import android.content.DialogInterface
import android.content.res.Configuration
import android.os.Bundle
import android.view.View
import android.widget.Button
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentActivity
import com.wa2c.android.prefs.Prefs

/**
 * Abstract dialog class.
 */
abstract class AbstractDialogFragment : DialogFragment() {

    /** The called activity.  */
    protected lateinit var context: FragmentActivity
    /** Prefs */
    protected lateinit var prefs: Prefs
    /** Click listener. */
    var clickListener: ((dialog: DialogInterface?, which: Int, bundle: Bundle?) -> Unit)? = null


    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        context = activity!!
        prefs = Prefs(context)
        return super.onCreateDialog(savedInstanceState)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        dialog?.cancel() // close on rotation
    }

    override fun onStart() {
        super.onStart()
        setDefaultListener(DialogInterface.BUTTON_POSITIVE)
        setDefaultListener(DialogInterface.BUTTON_NEGATIVE)
        setDefaultListener(DialogInterface.BUTTON_NEUTRAL)
    }

    override fun onStop() {
        super.onStop()
        dialog?.cancel()
    }

    /**
     * Set button event. Use this onStart or later.
     * @param which The button id.
     */
    private fun setDefaultListener(which: Int) {
        (dialog as AlertDialog?)?.getButton(which)?.setOnClickListener {
            invokeListener(which)
        }
    }

    /**
     * Invoke click listener.
     * @param which The button id.
     * @param close True if this dialog close.
     */
    protected open fun invokeListener(which: Int, bundle: Bundle? = null, close: Boolean = true) {
        clickListener?.invoke(dialog, which, bundle)
        if (close)
            dialog?.dismiss()
    }

    /**
     * Fragment tag.
     */
    private val fragmentTag: String by lazy { this.javaClass.name }

    /***
     * Show dialog.
     * @param activity A activity.
     */
    fun show(activity: FragmentActivity?) {
        if (activity == null)
            return
        val manager = activity.supportFragmentManager
        val fragment = manager.findFragmentByTag(fragmentTag) as? AbstractDialogFragment
        fragment?.dismiss()
        super.show(manager, fragmentTag)
    }

}
